### ISSUE TITLE
Release Planning: Grin v5.0.0

### ISSUE THREAD

**Issue by lehnberg** [mimblewimble/grin-pm#287]
Planning issue for version 5.0.0 of `grin` and `grin-wallet`, which is to support the scheduled network-wide upgrade occurring at block 1,048,320, around Jan 15, 2021.

#### Scope freeze
TBD

#### Target release date
Beta: TBD
Release candidate: TBD
Final release: TBD

#### Discussion
Planning discussion can be had asynchronously on this issue, or in development [sub-teams](https://grin.mw/governance#node-development-team) on [keybase](https://keybase.io/team/grincoin).

#### Definitions

**P1 - Critical
P2 - Important
P3 - Fix if time**

📝: awaiting specification
🛠: in progress
✅: merged
🔍: awaiting review

---
## Milestones
* Node: To be created once scope has frozen
* Wallet: To be created once scope has frozen

## Items

Make suggestions by commenting in issue, preferably with a motivation and a proposed priority.

| #  |  Priority | Team | Description | Owner | Reference | Status |
|---|---|:---|:---|:---|:---|---|
1 | P(1,2,3)? | Node/Wallet | What is the issue | Who's owning it? | Link to RFC, issue, or PR | 📝|

## Out of scope
| # |  Priority | Team | Description | Reference | Motivation
|---|---|:---|:---|:---|:---|
| | | | |

**Comment by MCM-Mike** [comment]
**Make it full TOR compatible:**

- run Grin-Node in TOR only mode
- make it ready to connect to other Tor Grin-Nodes
(tbd)

**Comment by antiochp** [comment]
@yeastplume @j01tz what's the context for (1) - is this "multi-party txs" in the sense of payment channels etc. or referring to something else?

**Comment by antiochp** [comment]
I'd like to propose adding the following for wallet - 

* "mostly" lock free transactions. https://forum.grin.mw/t/mostly-lock-free-transactions/7313

**Comment by antiochp** [comment]
And for node - 

* allow duplicate outputs

This one is presumably contingent on us deciding we want to do this.

**Comment by lehnberg** [comment]
@DavidBurkett:
> I think we should revisit block weight anyway before the final hardfork. I have some concerns with the current approach that are worth discussing. I'll add it to my list of forum posts I need to write

[putting as placeholder for myself to remember]

**Comment by antiochp** [comment]
> @DavidBurkett:
> 
> > I think we should revisit block weight anyway before the final hardfork. I have some concerns with the current approach that are worth discussing. I'll add it to my list of forum posts I need to write
> 
> [putting as placeholder for myself to remember]

Issue for "max block weight" discussion (related to above): https://github.com/mimblewimble/grin/issues/3368

Fee discussion also (loosely) related to block weight: https://github.com/mimblewimble/grin/issues/3369

**Comment by yeastplume** [comment]
* Consolidation of wallet commands / API functions (e.g no need for separate 'receive', 'finalize' etc commands) Doesn't have to be 5.0.0, but a 4.1.0 should contain the 'final' version of all wallet commands so everyone has time to adjust before 5.0.0
* Method of Slatepack address cycling
* More optimised SQL backend option for scalability needs (does not need any particular HF or major/minor release)
* (Nebulous thought) any changes/standards that might be desirable for an offline transaction relaying solution? Highly dependent on what direction we decide to take with this.

**Comment by antiochp** [comment]
"coinbase inputs" (we should not need to specify output features when _spending_)

**Comment by quentinlesceller** [comment]
- Update our fork of https://github.com/mimblewimble/rust-secp256k1-zkp with latest optimisation such as https://github.com/mimblewimble/rust-secp256k1-zkp/pull/71  (does not need any particular HF or major/minor release)

**Comment by antiochp** [comment]
Quick update on where we are currently with remaining items - 

* https://github.com/mimblewimble/grin-wallet/pull/532 "deprecate http(s)": approved and ready to merge (@jaspervdm to merge)
* https://github.com/mimblewimble/grin/pull/3481 "fix fees": merge conflicts to be resolved and final review
* https://github.com/mimblewimble/grin-wallet/pull/526 "fix fees wallet": merge conflicts need resolving
* https://github.com/mimblewimble/grin/pull/3503 "disable api v1": ready to merge (@quentinlesceller to merge)
* https://github.com/mimblewimble/grin/pull/3496 "PIBD p2p msgs": approved and ready to merge (@jaspervdm to merge)
* https://github.com/mimblewimble/grin/pull/3500 "PIBD_HIST capabilities": ready to merge once 3496 is in (@antiochp to merge)

**Comment by antiochp** [comment]
* v5.0.0-beta.1 (node) Release Notes: https://github.com/mimblewimble/grin/issues/3506 [WIP]
